ci/infra/testrunner/platforms/platform.py

```diff
@@ -1,24 +1,145 @@
+
+        if not os.path.isdir(self.conf.log_dir):
+            os.mkdir(self.conf.log_dir)
+            logger.info(f"Created log dir {self.conf.log_dir}")
+
+        for node_type in node_ips:
+            for ip_address in node_ips[node_type]:
+                node_log_dir = self._create_node_log_dir(ip_address, node_type, self.conf.log_dir)
+                logging_error = self.utils.collect_remote_logs(ip_address, self.logs, node_log_dir)
+
+                if logging_error:
+                    logging_errors = logging_error
+
+        platform_log_error = self._get_platform_logs()
+
+        if platform_log_error:
+            logging_errors = platform_log_error
+
+        return logging_errors
+
+    def get_lb_ipaddr(self):
+        """
+        Get the IP of the Load Balancer
+        :return:
+        """
+        pass
+
+    def get_nodes_ipaddrs(self, role):
+        """
+        Get the IP addresses of the given type of node
+        :param role: the type of node
+        :return:
+        """
+        return []
+
+    @step
+    def provision(self, num_master=-1, num_worker=-1, retries=4):
+        """Provision a cluster"""
+        if num_master > -1 or num_worker > -1:
+            logger.warning("Overriding number of nodes")
+            if num_master > -1:
+                self.conf.master.count = num_master
+                logger.warning("   Masters:{} ".format(num_master))
+
+            if num_worker > -1:
+                self.conf.worker.count = num_worker
+                logger.warning("   Workers:{} ".format(num_worker))
+
+        # TODO: define the number of retries as a configuration parameter
+        for i in range(0, retries):
+            retry = i + 1
+
+            try:
+                self._provision_platform()
+                break
+            except Exception as ex:
+                logger.warning(f"Provision attempt {retry}/{retries} failed")
+                if retry == retries:
+                    raise Exception(f"Failed {self.__class__.__name__} deployment") from ex
+
+    def ssh_run(self, role, nr, cmd):
+        ip_addrs = self.get_nodes_ipaddrs(role)
+        if nr >= len(ip_addrs):
+            raise ValueError(f'Node {role}-{nr} not deployed in platform')
+
+        self.utils.ssh_run(ip_addrs[nr], cmd)
+
+    @staticmethod
+    def _create_node_log_dir(ip_address, node_type, log_dir_path):
+        node_log_dir_path = os.path.join(log_dir_path, f"{node_type}_{ip_address.replace('.', '_')}")
+
+        if not os.path.isdir(node_log_dir_path):
+            os.mkdir(node_log_dir_path)
+            logger.info(f"Created log dir {node_log_dir_path}")
+
+        return node_log_dir_path
+
+    @staticmethod
+    def _cleanup_files(files):
+        for file in files:
+            if os.path.exists(file):
+                logger.info(f"Cleaning up {file}")
+
+                if os.path.isfile(file):
+                    os.remove(file)
+                else:
+                    shutil.rmtree(file)
```

Comment:
Is there any concern over what happens if a subsequent file in the `files` iterable was located beneath a directory removed by `rmtree`?  It might be a good idea to sort `files` to enforce processing leaf-first, but that's probably deferrable for a separate PR (if it's a potential problem to begin with).

Catching the relevant exception (and perhaps just emitting it as a logged warning) on failure to remove for whatever reason may also be good enough.  File not found - ignoreable.  Permission denied - maybe a problem. :)

Reply:
I think files that might of been deleted already should be caught by `if os.path.exists(file):`

Reply:
Oh yeah. Doh. :D